keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,365 @@ ensure all tenants have rights to the images that are already present on a host.
+      would already grant image access irregardless of this pull operation result for any next
+      pull attempts.
+
+The kubelet uses the following Golang interface to interact with the cache:
+```go
+// ImagePullManager keeps the state of images that were pulled and which are
+// currently still being pulled.
+// It should keep an internal state of images currently being pulled by the kubelet
+// in order to determine whether to destroy a "pulling" record should an image
+// pull fail.
+type ImagePullManager interface {
+    // RecordPullIntent records an intent to pull an image and should be called
+    // before a pull of the image occurs.
+    //
+    // RecordPullIntent() should be called before every image pull. Each call of
+    // RecordPullIntent() must match exactly one call of RecordImagePulled()/RecordImagePullFailed().
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordPullIntent(image string) error
+    // RecordImagePulled writes a record of an image being successfully pulled
+    // with ImagePullCredentials.
+    //
+    // `credentials` must not be nil and must contain either exactly one Kubernetes
+    // Secret coordinates in the `.KubernetesSecrets` slice or set `.NodePodsAccessible`
+    // to `true`.
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordImagePulled(image, imageRef string, credentials *imagemanager.ImagePullCredentials)
+    // RecordImagePullFailed should be called if an image failed to pull.
+    //
+    // Internally, it lowers its reference counter for the given image. If the
+    // counter reaches zero, the pull intent record for the image is removed.
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordImagePullFailed(image string)
+    // MustAttemptImagePull evaluates the policy for the image specified in
+    // `image` and if the policy demands verification, it checks the internal
+    // cache to see if there's a record of pulling the image with the presented
+    // set of credentials or if the image can be accessed by any of the node's pods.
+    //
+    // Returns true if the policy demands verification and no record of the pull
+    // was found in the cache.
+    //
+    // `image` is the content of the pod's container `image` field.
+    MustAttemptImagePull(image, imageRef string, credentials []imagemanager.ImagePullSecret) bool
+    // PruneUnknownRecords deletes all of the cache ImagePulledRecords for each of the images
+    // whose imageRef does not appear in the `imageList` iff such an record was last updated
+    // _before_ the `until` timestamp.
+    //
+    // This method is only expected to be called by the kubelet's image garbage collector.
+    // `until` is a timestamp created _before_ the `imageList` was requested from the CRI.
+    PruneUnknownRecords(imageList []string, until time.Time)
```

Comment:
does the timestamp here mean we need to keep track of the time when a given imagePulledRecord was created? is the plan to do this in a field inside the record, or with a filesystem timestamp, or in memory?

Reply:
I originally intended to have this be an implementation detail of the cache, which would imply the files' "modify" metadata information in the alpha case.
Should we explicitly track the `LastUpdatedTime metav1.Time` timestamp instead to avoid fs timestamp being unreliable?

Reply:
I think it's probably better to make the update time explicit in the file